### PR TITLE
Gen hash

### PR DESCRIPTION
**Pull request by EItanya**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1711

### CODE REVIEW

File: api/v1/metadata.proto [hunk 1]
```diff
@@ -26,24 +29,24 @@ message Metadata {
     // Cluster indicates the cluster this resource belongs to
     // Cluster is only applicable in certain contexts, e.g. Kubernetes
     // An empty string here refers to the local cluster
-    string cluster = 7;
+    string cluster = 7 [(extproto.skip_hashing) = true];
```
Comment (yuval-k) [inline api/v1/metadata.proto]: watch

File: api/v1/solo-kit.proto [hunk 2]
```diff
@@ -24,7 +24,7 @@ message Resource {
     // if set to true, changes in annotations will not cause a new snapshot to be emitted
     bool skip_hashing_annotations = 5;
 }
-
-extend google.protobuf.FieldOptions {
-    bool skip_hashing = 10000;
-}
+//
+//extend google.protobuf.FieldOptions {
```
Comment (yuval-k) [inline api/v1/solo-kit.proto]: goodbye

File: pkg/api/v1/resources/resource_interface.go [hunk 3]
```diff
@@ -404,7 +405,7 @@ func (list InputResourceList) AsResourceList() ResourceList {
 
 type HashableResource interface {
```
Comment (yuval-k) [inline pkg/api/v1/resources/resource_interface.go]: goodbye?

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go [hunk 4]
```diff
@@ -251,9 +254,16 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
 		snapshots <- &initialSnapshot
 
 		timer := time.NewTicker(time.Second * 1)
-		previousHash := currentSnapshot.Hash()
+		previousHash, err := currentSnapshot.Hash(nil)
+		if err != nil {
+			contextutils.LoggerFrom(ctx).DPanicw("error while hashing, this should never happen", zap.Error(err))
+		}
```
Comment (yuval-k) [inline pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: either panic on error
or return
inc a metric

File: pkg/code-generator/codegen/templates/snapshot_emitter_template.go [hunk 5]
```diff
@@ -251,9 +254,16 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
...
 		sync := func() {
-			currentHash := currentSnapshot.Hash()
+			currentHash, err := currentSnapshot.Hash(nil)
+			// this should never happen, so panic if it does
+			if err != nil {
+				contextutils.LoggerFrom(ctx).DPanicw("error while hashing, this should never happen", zap.Error(err))
```
Comment (yuval-k) [inline pkg/code-generator/codegen/templates/snapshot_emitter_template.go]: either panic on error
or return
inc a metric

File: pkg/code-generator/codegen/templates/snapshot_simple_emitter_template.go [hunk 6]
```diff
@@ -58,7 +60,10 @@ func (c *{{ lower_camel .GoName }}SimpleEmitter) Snapshots(ctx context.Context)
 		timer := time.NewTicker(time.Second * 1)
 		var previousHash uint64
 		sync := func() {
-			currentHash := currentSnapshot.Hash()
+			currentHash, err := currentSnapshot.Hash(nil)
+			if err != nil {
+				contextutils.LoggerFrom(ctx).DPanicw("error while hashing, this should never happen", zap.Error(err))
```
Comment (yuval-k) [inline pkg/code-generator/codegen/templates/snapshot_simple_emitter_template.go]: same (return or panic or stat)

File: pkg/code-generator/parser/parser_resource.go [hunk 7]
```diff
@@ -229,7 +229,7 @@ func describeResource(messageWrapper ProtoMessageWrapper) (*model.Resource, erro
 
 	hasStatus := hasField(msg, "status", statusTypeName)
 
-	fields := collectFields(msg)
+	// fields := collectFields(msg)
```
Comment (yuval-k) [inline pkg/code-generator/parser/parser_resource.go]: bye

File: test/mocks/tests/benchmark_test.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/mitchellh/hashstructure"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/solo-io/solo-kit/pkg/api/v1/resources/core"
+
+	// . "github.com/onsi/gomega"
+	v1 "github.com/solo-io/solo-kit/test/mocks/v1"
+)
+
+var _ = FDescribe("hashing", func() {
```
Comment (yuval-k) [inline test/mocks/tests/benchmark_test.go]: unf

File: test/mocks/tests/eventloop_test.go [hunk 9]
```diff
@@ -11,7 +11,7 @@ import (
 	v1 "github.com/solo-io/solo-kit/test/mocks/v1"
 )
 
-var _ = Describe("Eventloop", func() {
+var _ = FDescribe("Eventloop", func() {
```
Comment (yuval-k) [inline test/mocks/tests/eventloop_test.go]: unf

File: api/v1/metadata.proto [hunk 10]
```diff
@@ -39,11 +42,11 @@ message Metadata {
 
     // Annotations is an unstructured key value map stored with a resource that may be
     // set by external tools to store and retrieve arbitrary metadata.
-    map<string, string> annotations = 6;
+    map<string, string> annotations = 6 [(extproto.skip_hashing) = true];
```
Comment (ilackarms) [inline api/v1/metadata.proto]: rm

File: pkg/code-generator/codegen/templates/snapshot_template.go [hunk 11]
```diff
@@ -30,30 +33,43 @@ func (s {{ .GoName }}Snapshot) Clone() {{ .GoName }}Snapshot {
...
-
+	hasher := fnv.New64()
 {{- range .Resources}}
-	fields = append(fields, zap.Uint64("{{ lower_camel .PluralName }}", s.hash{{ upper_camel .PluralName }}() ))
+	{{ upper_camel .PluralName }}Hash, _ := s.hash{{ upper_camel .PluralName }}(hasher)
+	fields = append(fields, zap.Uint64("{{ lower_camel .PluralName }}", {{ upper_camel .PluralName }}Hash ))
 {{- end}}
-
-	return append(fields, zap.Uint64("snapshotHash",  s.Hash()))
+	snapshotHash, _ := s.Hash(hasher)
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/snapshot_template.go]: log error

**Pull request merged**